### PR TITLE
[corlib] Don't capture the execution context when UnsafeRegisterWaitForSingleObject is used. Fixes #60029

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

**Comment by marek-safar** [comment]
build

File: mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs [hunk 1]
```diff
@@ -216,5 +216,45 @@ void GetAvailableThreads_Callback (object state)
...
+				var_2 = asyncLocal.Value;
+				Interlocked.Increment (ref total);
+			}, null, 1, false);
+
+			ThreadPool.UnsafeRegisterWaitForSingleObject (evt, (state, to) => {
+				var_3 = asyncLocal.Value;
+				Interlocked.Increment (ref total);
+			}, null, 1, false);
+
+			while (total < 4)
```
Comment (marek-safar) [inline mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs]: I'd use a CountdownEvent with timeout as this block indefinitely on an error

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs [hunk 2]
```diff
@@ -1507,7 +1507,10 @@ bool               compressStack
                 throw new NotSupportedException ("Timeout is too big. Maximum is Int32.MaxValue");
 
             RegisteredWaitHandle waiter = new RegisteredWaitHandle (waitObject, callBack, state, new TimeSpan (0, 0, 0, 0, (int) millisecondsTimeOutInterval), executeOnlyOnce);
-            QueueUserWorkItem (new WaitCallback (waiter.Wait), null);
+			if (compressStack)
```
Comment (luhenry) [inline mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: This should be between `#if MONO`

File: mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs (hunk 1 above)
Comment (luhenry) [reply mcs/class/corlib/Test/System.Threading/ThreadPoolTest.cs]: I'd use `ManuelResetEvent` for each case, so you can figure out which one has finished and which one hasn't. You can wait on all of them with a `Assert.IsTrue(WaitHandle.WaitAll(mres, 5000))`.

File: mcs/class/referencesource/mscorlib/system/threading/threadpool.cs (hunk 2 above)
Comment (akoeplinger) [reply mcs/class/referencesource/mscorlib/system/threading/threadpool.cs]: It already is, see line 1471.

Fixing the indentation to match the surrounding code would be good though :)

**Comment by kumpera** [comment]
Fixed formating and moved to a CountdownEvent with a timed wait.

**Pull request merged**